app/solvers/exact_solver.py

```diff
@@ -105,63 +59,106 @@ def duplicate(items):
         # vec[1] = x, vec[2] = y, vec[3] = quantity
         c = City(k, vec[1], vec[2], vec[3])
-        c.getNeighbours(d)
-        cities_dict[k] = c
-    return cities_dict
+        c.getNeighbours(graph)
+        cities_dict_temp[k] = c
+    return cities_dict_temp
 
 
-def choose_the_best_path(lista, cities_list):
-    podsumowanie={}
+#resources is a list of possible paths with an eye to whole_time, cities list has info about "visited"  returns summary - paths with profits
+def choose_the_best_path(resources, cities_list):
+    summary ={}
     g = 0
-    i = 1
-    for j in lista:
-        zysk = 0
-        for krok in j:
-            if cities_list[krok].visited == False:
-                zysk = zysk + cities_list[krok].value
-                cities_list[krok].visited = True
-        podsumowanie[zysk] = j
+    for j in resources:
+        profit = 0
+        for stepOne in j:
+            if cities_list[stepOne].visited == False:
+                profit = profit + cities_list[stepOne].value
+                cities_list[stepOne].visited = True
+        summary[profit] = j
         g = g + 1
-        for krok in j:
-            cities_list[krok].visited = False
-        # print(i)
-        i = i + 1
-    return podsumowanie
+        for stepTwo in j:
+            cities_list[stepTwo].visited = False
+    # print(summary)
+    return summary
 
 
-def return_the_best_value(dict):
-    naj = max(dict.keys())
+# returns the best profit from dictionary which consist of paths with profits
+def return_the_best_value(dict_temp):
+    naj = max(dict_temp.keys())
     return naj
 
 
-def return_cost_of_the_best_path(dict, best):
+# return the cost on the best (from profit side) path
+def return_cost_of_the_best_path(graph, dict_temp_two, best):
     w = 1
-    koszt = 0
-    do = len(dict[best]) - 1
-    for x in dict[best][:do]:
-        koszt = koszt + dict[x][dict[best][w]]
+    cost = 0
+    do = len(dict_temp_two[best]) - 1
+    for x in dict_temp_two[best][:do]:
+        cost = cost + graph[x][dict_temp_two[best][w]]
         w = w + 1
-    return koszt
+    return cost
 
-
-def create_answer_for_path_creation(dict, best):
+# create special form of answer for path creation
+def create_answer_for_path_creation(dict_temp_three, best):
     answer_plot = []
-    do = len(dict[best]) -1
+    do = len(dict_temp_three[best]) -1
     for x in range(do):
-        answer_plot.append((dict[best][x], dict[best][x + 1]))
+        answer_plot.append((dict_temp_three[best][x], dict_temp_three[best][x + 1]))
     return answer_plot
+
+
+def solve(cities: pd.DataFrame, paths: pd.DataFrame, time: pd.DataFrame):
+    assert isinstance(cities, pd.DataFrame), 'Wrong data format!'
+    assert isinstance(paths, pd.DataFrame), 'Wrong data format!'
+    assert isinstance(time, pd.DataFrame), 'Wrong data format!'
+
+    # finds all paths from chosen start point with eye on possible time
+    def find_all_possible_paths(graph, start, time_left, path=[], price=0):
```

Comment:
Definiowanie funkcji w innej funkcji nie jest najlepszym patternem. Czemu akurat tak robisz?

Reply:
Bo potrzebuje zmiennej nonlocal - i konkretnie chce żeby tak właśnie to wyglądało, bo zmienna ma być czyszczona przed każdym nowym wywowałaniem funkcji, która jest rekursją (dlatego nie ma jak inaczej jej wyczyścić, bo przy następnej iteracji ma nie być czyszczona)